pkg/cloudprovider/provider/azure/create_delete_resources.go

```diff
@@ -0,0 +1,282 @@
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2018-04-01/compute"
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2018-04-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/golang/glog"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// deleteInterfacesByMachineUID will remove all network interfaces tagged with the specific machine's UID.
+// The machine has to be deleted or disassociated with the interfaces beforehand, since Azure won't allow
+// us to remove interfaces connected to a VM.
+func deleteInterfacesByMachineUID(ctx context.Context, c *config, machineUID types.UID) error {
+	ifClient, err := getInterfacesClient(c)
+	if err != nil {
+		return fmt.Errorf("failed to create interfaces client: %v", err)
+	}
+
+	list, err := ifClient.List(ctx, c.ResourceGroup)
+	if err != nil {
+		return fmt.Errorf("failed to list interfaces in resource group %q", c.ResourceGroup)
+	}
+
+	allInterfaces := make([]network.Interface, 0)
+
+	for {
+		allInterfaces = append(allInterfaces, list.Values()...)
+		if list.NotDone() {
+			list.Next()
+		} else {
+			break
+		}
+	}
+
+	for _, iface := range allInterfaces {
+		if iface.Tags != nil && *iface.Tags[machineUIDTag] == string(machineUID) {
+			future, err := ifClient.Delete(ctx, c.ResourceGroup, *iface.Name)
+			if err != nil {
+				return err
+			}
+
+			if err = future.WaitForCompletion(ctx, ifClient.Client); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// deleteIPAddressesByMachineUID will remove public IP addresses tagged with the specific machine's UID.
+// Their respective network interfaces have to be deleted or disassociated with the IPs beforehand, since
+// Azure won't allow us to remove IPs connected to NICs.
+func deleteIPAddressesByMachineUID(ctx context.Context, c *config, machineUID types.UID) error {
+	ipClient, err := getIPClient(c)
+	if err != nil {
+		return fmt.Errorf("failed to create IP addresses client: %v", err)
+	}
+
+	list, err := ipClient.List(ctx, c.ResourceGroup)
+	if err != nil {
+		return fmt.Errorf("failed to list public IP addresses in resource group %q", c.ResourceGroup)
+	}
+
+	allIPs := make([]network.PublicIPAddress, 0)
+
+	for {
+		allIPs = append(allIPs, list.Values()...)
+		if list.NotDone() {
+			list.Next()
+		} else {
+			break
+		}
+	}
+
+	for _, ip := range allIPs {
+		if ip.Tags != nil && ip.Tags[machineUIDTag] != nil && *ip.Tags[machineUIDTag] == string(machineUID) {
+			future, err := ipClient.Delete(ctx, c.ResourceGroup, *ip.Name)
+			if err != nil {
+				return err
+			}
+
+			if err = future.WaitForCompletion(ctx, ipClient.Client); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func deleteVMsByMachineUID(ctx context.Context, c *config, machineUID types.UID) error {
+	vmClient, err := getVMClient(c)
+	if err != nil {
+		return err
+	}
+
+	list, err := vmClient.ListAll(ctx)
+	if err != nil {
+		return err
+	}
+
+	allServers := make([]compute.VirtualMachine, 0)
```

Comment:
I think its more idiomatic to write `var allServers []compute.VirtualMachine` instead of instantiating a zero-lenght array

Reply:
fixed